src/com/sap/piper/DefaultValueCache.groovy

```diff
@@ -0,0 +1,27 @@
+package com.sap.piper
+
+class DefaultValueCache implements Serializable {
+    private static DefaultValueCache instance
+
+    private Map defaultValues
```

Comment:
I guess as Daniel put it, it is more Groovy like and should be ok

Reply:
We could discuss whether we should omit Map here. I would not use generics as you try to minimize using types in groovy. Furthermore, I would not be really helpful, as it would be <String, Object> I guess.

Reply:
I would like to explain the reason for my question above. It is related to getDefaultValues below which returns the map.

The key question is: would we like to be sure that the default values are really what we configured accordingly in the persistency layer? If we say: "well, if somebody in a pipeline deliberately changes the default values from the persistency, it is OK for us ..." we can keep everything like it is. But if we say: "The default values must remain like they are configured in the persistency layer" then we have an issue with the current code. Again: I'm fine with the first statement, but we should be very clear about what we want.

When handing over the map to outside the caller of getDefaultValues can change the map, full CRUD. In case we would like to be sure about what is defined in persistency we can
1.) create a defensive copy. With that the caller cannot modify the defaults in the class, of course he can still modify his copy. Of course it needs to be a deep copy: a value in the defensive copy != corresponding value in the original (only important if the values are mutable).
2.) Create a unmodifiable map. With this the caller of the method can only read. See java.utils.Collections. But here we have the link to the generics here: This only makes sense if the values itself are immutable. Otherwise a caller cannot change the map, but the values contained in the map. with e.g. string we would be immutable. Of course it would be possible to create the unmodifiable map already in the constructor.

Having said all this I repeat what I already commented in the beginning: In case the approach with the default values does not mean that the defaults cannot be modified in the pipeline we can keep everything like it is.

Reply:
I think, it is not a problem if someone modifies the default values in the pipeline. The Idea here was just to maintain these default values at a central place and not in the step.